### ISSUE TITLE
Android: Biometric Prompt not showing again

### ISSUE THREAD

**Issue by AhmedFarazQaimkhani** [oblador/react-native-keychain#504]
Biometric Prompt not showing again when click on login via biometric after unlocking device while on Login Screen. It logs in without showing prompt.
<img width="265" alt="Screenshot 2021-09-13 at 6 34 01 PM" src="https://user-images.githubusercontent.com/50355228/133093095-23956ce7-1ce5-4f3e-b0d2-6ad453dd11e7.png">

**Comment by anhquan291** [comment]
`  await Keychain.setGenericPassword(username, password, {
      accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY,
      accessible: Keychain.ACCESSIBLE.ALWAYS,
      authenticationType: Keychain.AUTHENTICATION_TYPE.BIOMETRICS,
      storage: Keychain.STORAGE_TYPE.RSA,
    });`

Facing the same issue here. Any solution, please? Thanks a lot

**Comment by AlvaroBrasilia** [comment]
Try to insert `accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY` argument in options for Keychain.getGenericPassword(options);

async load() {
    try {
      const options = {
        authenticationPrompt: {
          title: 'Authentication needed',
          subtitle: 'Subtitle',
          description: 'Some descriptive text',
          cancel: 'Cancel',
        },
        **accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY,**
        service: 'MAP',
      };
      const credentials = await Keychain.getGenericPassword(options);
      console.log('credentials', credentials);
      if (credentials) {
        this.setState({ ...credentials, status: 'Credentials loaded!' });
      } else {
        this.setState({ status: 'No credentials stored.' });
      }
    } catch (err) {
      this.setState({ status: 'Could not load credentials. ' + err });
    }
  }

**Comment by anhquan291** [comment]
> Try to insert `accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY` argument in options for Keychain.getGenericPassword(options);
> 
> async load() {
> try {
> const options = {
> authenticationPrompt: {
> title: 'Authentication needed',
> subtitle: 'Subtitle',
> description: 'Some descriptive text',
> cancel: 'Cancel',
> },
> **accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY,**
> service: 'MAP',
> };
> const credentials = await Keychain.getGenericPassword(options);
> console.log('credentials', credentials);
> if (credentials) {
> this.setState({ ...credentials, status: 'Credentials loaded!' });
> } else {
> this.setState({ status: 'No credentials stored.' });
> }
> } catch (err) {
> this.setState({ status: 'Could not load credentials. ' + err });
> }
> }

Hi @AlvaroBrasilia , thanks for your answer. 

Unfortunately,  it doesn't work. The Fingerprint prompt showed only once the first time. After that still can access the data without asking for Fingerprint. Any helps, please? Thank you so much

**Comment by AlvaroBrasilia** [comment]
@anhquan291 did you set `sevice` in `Keychain.setGenericPassword()` and `Keychain.getGenericPassword()`?

**Comment by AhmedFarazQaimkhani** [comment]
> @anhquan291 did you set `sevice` in `Keychain.setGenericPassword()` and `Keychain.getGenericPassword()`?

Hi @AlvaroBrasilia, thanks for your response. 
I'm doing the same, but still facing this issue.. I did set service in Keychain.setGenericPassword() and Keychain.getGenericPassword()

**Comment by AhmedFarazQaimkhani** [comment]
@AlvaroBrasilia here is me get and set method
Keychain.getGenericPassword({
        service: serviceName,
        storage: Keychain.STORAGE_TYPE.RSA,
        accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY,
        accessible: Keychain.ACCESSIBLE.ALWAYS,
        authenticationType: Keychain.AUTHENTICATION_TYPE.BIOMETRICS,
)}

Keychain.setGenericPassword(email, password, {
        service: serviceName,
        accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY,
        accessible: Keychain.ACCESSIBLE.ALWAYS,
        authenticationType: Keychain.AUTHENTICATION_TYPE.BIOMETRICS,
        storage: Keychain.STORAGE_TYPE.RSA,
      });

**Comment by anhquan291** [comment]
> @anhquan291 did you set `sevice` in `Keychain.setGenericPassword()` and `Keychain.getGenericPassword()`?

Yes, I do. The Prompt up shows again after the 2 times getting user data. It seems that on Android you just need to unlock for the first time. Thank you

**Comment by jess-dovetail** [comment]
This is what I have and it seems to work fine on both platforms
```
await Keychain.setInternetCredentials(
      SOMEID,
      email,
      password,
      {
        accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_CURRENT_SET,
        accessible: Keychain.ACCESSIBLE.WHEN_UNLOCKED_THIS_DEVICE_ONLY,
      }
    );
```

**Comment by AlvaroBrasilia** [comment]
Hello, I think we have a security issue here.

The issue is caused by

[setUserAuthenticationValidityDurationSeconds()](https://github.com/oblador/react-native-keychain/blob/67bee0739cabfe931cf048c409266b0be44693f3/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreRsaEcb.java#L229)

The authentication is valid for the defined time in seconds.

If you unlock the device and click on load button when authentication time is validity the biometry modal don't show and the values are definied without a new authentication request.

Please, click to see the video below.

[![Watch the video](https://user-images.githubusercontent.com/2683945/133779116-8753e25e-1f0f-4913-873e-97c16f09361f.png)](https://www.youtube.com/watch?v=CmzoovDK28s)

**Comment by anhquan291** [comment]
I got it. It is really helpful. Thank you so much.

**Comment by pabloearg** [comment]
> Hello, I think we have a security issue here.
> 
> The issue is caused by
> 
> [setUserAuthenticationValidityDurationSeconds()](https://github.com/oblador/react-native-keychain/blob/67bee0739cabfe931cf048c409266b0be44693f3/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreRsaEcb.java#L229)
> 
> The authentication is valid for the defined time in seconds.
> 
> If you unlock the device and click on load button when authentication time is validity the biometry modal don't show and the values are definied without a new authentication request.
> 
> Please, click to see the video below.
> 
> [![Watch the video](https://user-images.githubusercontent.com/2683945/133779116-8753e25e-1f0f-4913-873e-97c16f09361f.png)](https://www.youtube.com/watch?v=CmzoovDK28s)

tested on an android 9 emulator, compiled with API 30, and it didn't work

[setUserAuthenticationValidityDurationSeconds()](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder#setUserAuthenticationValidityDurationSeconds(int)
) was deprecated in API 30. 
Maybe that's the reason why it doesn't work now?

**Comment by CaioQuirinoMedeiros** [comment]
> Hello, I think we have a security issue here.
> 
> The issue is caused by
> 
> [setUserAuthenticationValidityDurationSeconds()](https://github.com/oblador/react-native-keychain/blob/67bee0739cabfe931cf048c409266b0be44693f3/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreRsaEcb.java#L229)


I'm afraid this was a fix for another problem reported on #318. Fixed on #339
😟

**Comment by tylim88** [comment]
I am having the same issue, the biometric(fingerprint) or pattern or whatever authentication are not working on my android phone no matter what I try, I gave up and use [expo local authentication](https://docs.expo.dev/versions/latest/sdk/local-authentication/#localauthenticationoptions) to handle the authentication part instead.

**Comment by sgal** [comment]
The issue here is indeed with the validity parameter on key spec. Current implementation only exposes generic API that unlocks entire Keystore if the user recently passed strong crypto challenge (passcode, Class 3 biometric, pattern, etc.).

Android has API to authorize a specific operation with Keystore, which in theory would help in this use case. But the library does not support it at the moment.

**Comment by mpatafio** [comment]
Probably related: https://github.com/oblador/react-native-keychain/issues/336

**Comment by saitejas-cypherd** [comment]
**This is what I have. Still not able to get the authentication prompt. Please kindly help me with this**


await Keychain.setInternetCredentials(
      SOMEID,
      email,
      password,
      {
        accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_CURRENT_SET_OR_DEVICE_PASSCODE,
        accessible: Keychain.ACCESSIBLE.WHEN_UNLOCKED_THIS_DEVICE_ONLY,
       storage: Keychain.STORAGE_TYPE.AES
      });

await Keychain.getInternetCredentials(
        SOMEID,
        {
          authenticationPrompt: {
            title: "<Prompt text>",
          },
        }
      );

**Comment by clauderobi** [comment]
I am getting the prompt on Android but only if the STORAGE_TYPE is RSA. Why not AES?

**Comment by vaivai130** [comment]
^Bump, anyone make any headway on this?

**Comment by MiracleUFO** [comment]
What's the update on this?
